### ISSUE TITLE
Chore/fix: `ModelSaver` save-path collision for SeedVR2, then add its tiny test

### ISSUE THREAD

**Issue by ianscrivener** [mflux-community/mflux#621]
Another tweak arising from **tiny test** work

---
# PRD: Fix `ModelSaver` save-path collision for SeedVR2, then add its tiny test

## Background

`SeedVR2WeightDefinition3B`/`SeedVR2WeightDefinition7B`
(`src/mflux/models/seedvr2/weights/seedvr2_weight_definition.py`) each define two
components, `transformer` and `vae`, and give **both** the same `hf_subdir="."`:

```python
class SeedVR2WeightDefinition3B:
    @staticmethod
    def get_components() -> List[ComponentDefinition]:
        return [
            ComponentDefinition(
                name="transformer",
                hf_subdir=".",
                num_blocks=32,
                ...
                weight_files=["seedvr2_ema_3b_fp16.safetensors"],
            ),
            ComponentDefinition(
                name="vae",
                hf_subdir=".",
                num_blocks=4,
                ...
                weight_files=["ema_vae_fp16.safetensors"],
            ),
        ]
```

This is correct for **loading**: the real SeedVR2 HF repo has both `.safetensors`
files sitting flat at repo root, and `weight_files` disambiguates them by exact
filename on the load path.

It is **not safe for saving**. `ModelSaver._save_weights`
(`src/mflux/models/common/weights/saving/model_saver.py`) ignores `weight_files`
entirely — it only ever knows about `hf_subdir` (called `subdir` in that function) —
and always writes shard files as plain sequential names:

```python
@staticmethod
def _save_weights(base_path: str, bits: int, model: nn.Module, subdir: str) -> None:
    path = Path(base_path) / subdir
    path.mkdir(parents=True, exist_ok=True)
    weights = dict(tree_flatten(model.parameters()))
    shards = ModelSaver._split_weights(weights)
    for i, shard in enumerate(shards):
        shard_filename = f"{i}.safetensors"
        mx.save_safetensors(str(path / shard_filename), shard, {...})
        ...
    # writes path / "model.safetensors.index.json"
```

With both `transformer` and `vae` resolving `path` to the same directory
(`base_path / "."`), saving the second component overwrites the first's
`0.safetensors`, `1.safetensors`, ... and its `model.safetensors.index.json`.

Confirmed empirically (2026-08-17) via a standalone repro: saving tiny `transformer` +
`vae` components left exactly one `0.safetensors` in the output directory, and
reloading reported `transformer: key sets differ` — the transformer component
actually got the VAE's data back on reload. Also confirmed structurally: SeedVR2 is
the **only** weight definition in the repo where two non-tokenizer components share an
`hf_subdir` — every other model gives each component a distinct subdir (e.g.
`z_image`'s `ZImageWeightDefinition` uses `hf_subdir="vae"` / `"transformer"` /
`"text_encoder"` / `"tokenizer"`; same pattern in `ernie_image`, and everywhere else).

## Why this hasn't broken anything yet

`grep -rn "save_model\|ModelSaver" src/mflux/models/seedvr2/` returns nothing — no
SeedVR2 variant class defines `save_model`, and there is no `mflux-save-*` CLI entry
point wired up for SeedVR2 either. The collision is real but currently **latent**:
nothing in the shipped CLI can trigger `ModelSaver.save_model` with a SeedVR2 weight
definition today. It becomes a silent-corruption bug the moment someone adds save
support for SeedVR2 (e.g. quantized-model saving, matching `mflux-save` for other
models) without independently noticing and fixing this.

This was discovered as a side effect of attempting a tiny model-saving test for
SeedVR2 (see `.claude/notes/tiny-tests.md`, "SeedVR2 — reclassified" section) — the
test's save/reload roundtrip exercises exactly the path that would corrupt a real
save, which read-only dimension analysis alone didn't catch.

## Goal

1. Fix `ModelSaver` so two components sharing an `hf_subdir` (a legitimate load-side
   layout) don't collide on save. Two candidate approaches — pick one:
   - **(a) Prefix shard filenames by component name** in `_save_weights`/
     `_split_weights`, e.g. `f"{component_name}.{i}.safetensors"` instead of bare
     `f"{i}.safetensors"`, and likewise namespace `model.safetensors.index.json`
     per component (or merge weight_maps into one index file per subdir, keyed
     unambiguously). Minimal, doesn't touch `hf_subdir` semantics at all.
   - **(b) Give each component an independent save-time subdirectory**, decoupled
     from `hf_subdir` (which stays as the *load*-side layout hint). E.g. add a
     `save_subdir` field to `ComponentDefinition` (defaulting to `hf_subdir` for
     every other model, where it's already unique) that `ModelSaver` uses instead.
   Check whether either approach has knock-on effects on `WeightLoader`'s
   `_try_load_mflux_format` path (the round-trip re-load of an mflux-saved
   checkpoint) before committing to one — that path currently assumes the
   `hf_subdir`-based layout it wrote.
2. Confirm the fix doesn't change save/load behavior for every other model (all of
   which already have unique per-component `hf_subdir`s, so the fix should be a
   no-op for them — verify, don't just assume).
3. Once (1)-(2) land, write `tests/model_saving/test_tiny_model_saving_seedvr2.py`
   following the `mflux-model-tiny-test` skill. The dimension/shrinkability side was
   already fully worked out and is not blocked — see the specific tiny dimensions
   below.

## Non-goals

- Not adding `mflux-save` / a CLI entry point for SeedVR2 — this PRD only fixes the
  underlying collision so that *whenever* save support is added (here or elsewhere),
  it isn't silently broken. Adding the CLI surface is a separate decision.
- Not touching the FLUX.1/FIBO shared-class parameterization
  (`prd-parameterize-flux-timestep-norm.md`) or Depth Pro
  (`prd-parameterize-depth-pro.md`) — unrelated bugs in unrelated models.
- Not revisiting `weight_files`' role on the load side — it's correct as-is and
  unaffected by this fix.

## Already-solved: tiny-test dimensions (no further investigation needed)

Confirmed 2026-08-17 that every class under `seedvr2/model/` (`SeedVR2Transformer`,
`SeedVR2VAE`, `TransformerBlock`, `MMAttention`, `SwiGLUMLP`, `TimeEmbedding`,
`PatchIn`/`PatchOut`, `ResnetBlock3D`, `Attention3D`, `DownBlock3D`/`UpBlock3D`) takes
explicit dimension kwargs — this part was never blocked. For when the `ModelSaver`
fix lands:

- SeedVR2's real VAE is itself shrinkable (`block_out_channels`, `latent_channels`,
  `layers_per_block`) — use a real tiny `SeedVR2VAE`, `TinyVAEStandIn` isn't needed.
- Use `PatchIn(vid_in_channels=16, ...)`, not the default `33` — `33` produces
  `Linear(132, dim)` and 132 isn't a multiple of 64, so the model's own
  `quantization_predicate` would silently skip quantizing it.
- VAE `block_out_channels` must be a multiple of 32 (hardcoded
  `GroupNorm(num_groups=32)`) *and* 64 (quantization) — `(64, 64)` satisfies both.
- No `quantization_group_size` override (defaults to 64); no `skip_quantization`
  flags (uses `quantization_predicate` instead, which the harness already applies
  symmetrically — not the known krea2/qwen gap).
- Test only `SeedVR2WeightDefinition3B` (or whatever `resolve()` defaults to) — one
  test covers both 3B/7B, since `WeightLoader._load_component` tries
  `_try_load_mflux_format` first and returns on the tiny test's own saved metadata
  before `num_blocks` is ever consulted.

## Verification plan

Unlike the FLUX.1/FIBO and Depth Pro PRDs, this is a **test-infrastructure /
save-path bug fix**, not a change to a model's forward-pass math — so it does not
need the `mflux-debugging` export-then-compare numerical parity workflow. What it
does need:

1. Unit-level: after the fix, save a tiny `transformer` + `vae` pair (or reuse
   whatever repro script surfaced the bug) and confirm the output directory contains
   correctly separated files for both components — no filename collision, both
   reload with matching key sets and values.
2. Confirm no other model's save output changes at all — every other weight
   definition already uses unique `hf_subdir`s, so a diff of saved output
   layout/filenames for e.g. `z_image` or `ernie_image` before/after should be empty.
3. If SeedVR2 save support exists by the time this is picked up, run one real (or
   tiny) save + reload end-to-end and confirm both components come back correctly.
4. `just test-fast` / `just test` green.

## Acceptance criteria

- [ ] `ModelSaver` (or `ComponentDefinition`) changed so components sharing an
      `hf_subdir` no longer collide on save.
- [ ] Confirmed no behavior change for models with already-unique `hf_subdir`s.
- [ ] `tests/model_saving/test_tiny_model_saving_seedvr2.py` written and green, using
      the dimensions already worked out above.
- [ ] `.claude/notes/tiny-tests.md` coverage matrix updated: `seedvr2` from "❌
      blocked" to "✅".
- [ ] Consider whether `WeightLoader`'s `_try_load_mflux_format` needs a matching
      change, or whether it already tolerates the new save layout unchanged (it reads
      `model.safetensors.index.json` and `weight_map`, which any of the fix
      approaches above should keep intact — verify, don't assume).

## References

- `.claude/notes/tiny-tests.md` — "SeedVR2 — reclassified: blocked by a real
  ModelSaver bug, not dimensions" section (full investigation + repro details).
- `.cursor/skills/mflux-model-tiny-test/SKILL.md` — tiny test pattern.
- `src/mflux/models/common/weights/saving/model_saver.py` — `_save_weights`, the
  function that needs to change.
- `src/mflux/models/seedvr2/weights/seedvr2_weight_definition.py` — both
  `SeedVR2WeightDefinition3B` and `...7B` need their `ComponentDefinition`s to keep
  working after the fix (or gain a new field, per approach (b) above).
- `src/mflux/models/common/weights/loading/weight_definition.py` — `ComponentDefinition`
  dataclass, if approach (b) (a new `save_subdir` field) is chosen.
- Sibling PRDs for the same "investigation said trivial, implementation found a real
  bug" pattern: `prd-parameterize-flux-timestep-norm.md` (FLUX.1/FIBO),
  `prd-parameterize-depth-pro.md` (Depth Pro) — different root causes, same overall
  shape of "someone needs to scope and verify a `src/` fix before the tiny test can
  land."
